### PR TITLE
Fix `ckb-hash` depends `blake2b-ref` multiple times

### PR DESCRIPTION
**Pull request by eval-exec**
### What problem does this PR solve?

Problem Summary:
In latest develop branch, we found this:
```bash
$ cargo metadata --format-version 1 --all-features --manifest-path ./Cargo.toml
error: the crate `ckb-hash v0.112.0-pre (/home/exec/Projects/github.com/nervosnetwork/ckb/util/hash)` depends on crate `blake2b-ref v0.3.1` multiple times with different names
```
This PR want to fix this error.

### What is changed and how it works?

What's Changed:

### Related changes

- Remove `ckb-hash/default` feature, and make this feature's behavior as default

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code ci-runs-only: [ quick_checks,linters ]

Side effects

- None

### Release note 

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

**Comment by chenyukang** [comment]
can we add a command line like this in CI to avoid this issue?
```console
cargo metadata --format-version 1 --manifest-path /Users/yukang/code/ckb/./Cargo.toml --filter-platform aarch64-apple-darwin```

`aarch64-apple-darwin` is replaced the the CI host.

**Comment by eval-exec** [comment]
> can we add a command line like this in CI to avoid this issue?
> 
> ```
> cargo metadata --format-version 1 --manifest-path /Users/yukang/code/ckb/./Cargo.toml --filter-platform aarch64-apple-darwin```
> 
> `aarch64-apple-darwin` is replaced th CI host.
> ```

Sure

**Comment by eval-exec** [comment]
Waiting the CI jobs results.

File: util/gen-types/Cargo.toml [hunk 1]
```diff
@@ -20,7 +20,7 @@ check-data = []
 # Enable the `serialized-size` extension for CKB contract development in `no-std` env
 serialized-size = ["calc-hash"]
 # Enable all in `std` env
-std = ["molecule/std", "ckb-hash/default", "ckb-fixed-hash", "ckb-error", "ckb-occupied-capacity", "numext-fixed-uint"]
+std = ["molecule/std", "ckb-hash", "ckb-fixed-hash", "ckb-error", "ckb-occupied-capacity", "numext-fixed-uint"]
```
Comment (EthanYuan) [inline util/gen-types/Cargo.toml:23]: can remove “default-features = false“
in
ckb-hash = { path = "../hash", version = "= 0.112.0-pre", default-features = false, optional = true }

File: util/hash/Cargo.toml [hunk 2]
```diff
@@ -9,14 +9,8 @@ homepage = "https://github.com/nervosnetwork/ckb"
...
-
-[target.'cfg(target_arch = "wasm32")'.dependencies]
-blake2b = { package = "blake2b-ref", version = "0.3", optional = true }
+ckb-contract = [] # This feature is used for CKB contract development
 
 [dependencies]
-blake2b-ref = { version = "0.3", optional = true }
\ No newline at end of file
+blake2b-ref = { version = "0.3"}
+blake2b-rs = { version = "0.2", optional = true }
```
Comment (EthanYuan) [inline util/hash/Cargo.toml:16]:
```suggestion
blake2b-rs = { version = "0.2" }
```

File: util/hash/src/lib.rs [hunk 3]
```diff
@@ -9,11 +9,15 @@
 
 #![no_std]
 
-#[cfg(feature = "default")]
-pub use blake2b::{Blake2b, Blake2bBuilder};
 #[cfg(feature = "ckb-contract")]
 pub use blake2b_ref::{Blake2b, Blake2bBuilder};
 
+#[cfg(all(not(feature = "ckb-contract"), target_arch = "wasm32"))]
+pub use blake2b_rs::{Blake2b, Blake2bBuilder};
```
Comment (EthanYuan) [inline util/hash/src/lib.rs:16]:
```suggestion
pub use blake2b_ref::{Blake2b, Blake2bBuilder};
```

File: util/hash/src/lib.rs [hunk 4]
```diff
@@ -9,11 +9,15 @@
...
+
+#[cfg(all(not(feature = "ckb-contract"), not(target_arch = "wasm32")))]
+pub use blake2b_ref::{Blake2b, Blake2bBuilder};
```
Comment (EthanYuan) [inline util/hash/src/lib.rs:19]:
```suggestion
pub use blake2b_rs::{Blake2b, Blake2bBuilder};
```

File: util/hash/Cargo.toml (hunk 2 above)
Comment (EthanYuan) [reply util/hash/Cargo.toml:16]: The following code modification will need to be removed here.

**Comment by EthanYuan** [comment]
When I replace the following in my Cargo.toml file on this [branch](https://github.com/EthanYuan/ckb-std/tree/update) in [pr](https://github.com/nervosnetwork/ckb-std/pull/71)

```toml
ckb-types = { package = "ckb-gen-types", git = "https://github.com/EthanYuan/ckb.git", branch = "develop", default-features = false, optional = true }
```
with
```toml
ckb-types = { package = "ckb-gen-types", git = "https://github.com/eval-exec/ckb.git", branch = "exec/fix-cargo-metadata", default-features = false, optional = true }
```
I encounter an error when running the command in `ckb-std` ci `build-with-clang.yml`:

```yml
cargo build --verbose --target=riscv64imac-unknown-none-elf --features=build-with-clang
```

```bash
error: failed to run custom build command for `blake2b-rs v0.2.0`

Caused by:
  process didn't exit successfully: `/home/ethan/projects/cryptape/ckb-std/target/debug/build/blake2b-rs-5b354084b5d2fdc8/build-script-build` (exit status: 1)
  --- stdout
  TARGET = Some("riscv64imac-unknown-none-elf")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CC_riscv64imac-unknown-none-elf
  CC_riscv64imac-unknown-none-elf = None
  cargo:rerun-if-env-changed=CC_riscv64imac_unknown_none_elf
  CC_riscv64imac_unknown_none_elf = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  RUSTC_LINKER = None
  cargo:rerun-if-env-changed=CROSS_COMPILE
  CROSS_COMPILE = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = None
  cargo:rerun-if-env-changed=CFLAGS_riscv64imac-unknown-none-elf
  CFLAGS_riscv64imac-unknown-none-elf = None
  cargo:rerun-if-env-changed=CFLAGS_riscv64imac_unknown_none_elf
  CFLAGS_riscv64imac_unknown_none_elf = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  running: "riscv64-unknown-elf-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-march=rv64imac" "-mabi=lp64" "-mcmodel=medany" "-Wall" "-Wextra" "-o" "/home/ethan/projects/cryptape/ckb-std/target/riscv64imac-unknown-none-elf/debug/build/blake2b-rs-52d53686f90fa016/out/BLAKE2/ref/blake2b-ref.o" "-c" "BLAKE2/ref/blake2b-ref.c"

  --- stderr


  error occurred: Failed to find tool. Is `riscv64-unknown-elf-gcc` installed?
```

It looks like in a no-std environment, the issue is related to blake2b-rs.

**Comment by eval-exec** [comment]
Specify `"rust-analyzer.cargo.features": []` on rust-analyzer avoid this issue.

**Comment by eval-exec** [comment]
> When I replace the following in my Cargo.toml file on this [branch](https://github.com/EthanYuan/ckb-std/tree/update) in [pr](https://github.com/nervosnetwork/ckb-std/pull/71)
> 
> ```toml
> ckb-types = { package = "ckb-gen-types", git = "https://github.com/EthanYuan/ckb.git", branch = "develop", default-features = false, optional = true }
> ```
> 
> with
> 
> ```toml
> ckb-types = { package = "ckb-gen-types", git = "https://github.com/eval-exec/ckb.git", branch = "exec/fix-cargo-metadata", default-features = false, optional = true }
> ```
> 
> I encounter an error when running the command in `ckb-std` ci `build-with-clang.yml`:

`ckb-std` always need `blake2b-ref`, I think you should specify `ckb-contract` feature in `ckb-std`.

**Pull request closed without merging**